### PR TITLE
Parser: fix wrong/missing locations of various AST nodes

### PR DESCRIPTION
**Pull request by FnControlOption**
I've included examples in tests and commit description, but if you want I can split this PR into multiple PRs.

### CODE REVIEW

File: spec/compiler/parser/parser_spec.cr [hunk 1]
```diff
@@ -22,13 +22,32 @@ private def it_parses(string, expected_node, file = __FILE__, line = __LINE__)
...
   it "gets corrects end location for #{source.inspect}", file, line do
-    parser = Parser.new("#{source}; 1")
+    string = "#{source}; 1"
+    parser = Parser.new(string)
     node = parser.parse.as(Expressions).expressions[0]
+    loc = node.location.not_nil!
     end_loc = node.end_location.not_nil!
     end_loc.line_number.should eq(line_number)
     end_loc.column_number.should eq(column_number)
+    source_between(string, loc, end_loc).should eq(source)
```
Comment (straight-shoota) [inline spec/compiler/parser/parser_spec.cr:50]: What's the purpose of this expectation? I don't see how the source string comparison adds any relevant information to just the end location.

Comment (FnControlOption) [reply spec/compiler/parser/parser_spec.cr:50]: It’s useful for checking if the start location is set correctly (so maybe I should move this to a separate test). However, I think the string comparison is easier to understand at a glance than line/column numbers (when the test fails), so perhaps we can delete the number comparisons altogether?

**Comment by HertzDevil** [comment]
Where does the internal compiler error come from?

**Comment by Sija** [comment]
This would be a great help to a tools like ameba, is there anything holding this back from being merged?

File: src/compiler/crystal/syntax/parser.cr [hunk 2]
```diff
@@ -4789,31 +4826,25 @@ module Crystal
...
-        next_token
-        var_out
+        var = Underscore.new
       else
         raise "expecting variable or instance variable after out"
       end
+      var.location = @token.location
+      var.end_location = token_end_location
+      next_token
+      Out.new(var).at(location).at_end(var)
```
Comment (straight-shoota) [inline src/compiler/crystal/syntax/parser.cr:4847]:
```suggestion
      Out.new(var).at(var)
```

File: src/compiler/crystal/syntax/parser.cr [hunk 3]
```diff
@@ -4789,31 +4826,25 @@ module Crystal
...
-        underscore = Underscore.new.at(location)
-        var_out = Out.new(underscore).at(location)
-        next_token
-        var_out
+        var = Underscore.new
       else
         raise "expecting variable or instance variable after out"
       end
+      var.location = @token.location
+      var.end_location = token_end_location
```
Comment (straight-shoota) [inline src/compiler/crystal/syntax/parser.cr:4845]: This is a bit more concise:
```suggestion
      var.at(@token.location).at_end(token_end_location)
```

**Review by straight-shoota — APPROVED** [review]

File: src/compiler/crystal/syntax/parser.cr (hunk 2 above)
Comment (straight-shoota) [reply src/compiler/crystal/syntax/parser.cr:4836]: Sorry, I missed that 🙇

**Comment by Sija** [comment]
Could this make it into 1.7.0, pretty please? :)

**Comment by Sija** [comment]
CI is 💚 , 2nd review anyone?

**Review by asterite — APPROVED** [review]

**Pull request merged**